packages/edit-site/src/components/list/actions/rename-menu-item.js

```diff
@@ -0,0 +1,92 @@
+/**
+ * WordPress dependencies
+ */
+import { useState } from '@wordpress/element';
+import { useDispatch } from '@wordpress/data';
+import { useEntityProp, store as coreStore } from '@wordpress/core-data';
+import { __ } from '@wordpress/i18n';
+import {
+	Button,
+	Flex,
+	FlexItem,
+	MenuItem,
+	Modal,
+	TextControl,
+} from '@wordpress/components';
+
+export default function RenameMenuItem( { template, onClose } ) {
+	const [ isModalOpen, setIsModalOpen ] = useState( false );
+
+	const { saveEditedEntityRecord } = useDispatch( coreStore );
+	const [ title, setTitle ] = useEntityProp(
+		'postType',
+		template.type,
+		'title',
+		template.id
+	);
+
+	async function onTemplateRename( event ) {
+		event.preventDefault();
+		setIsModalOpen( false );
+		onClose();
+
+		// Persist edited entity.
+		await saveEditedEntityRecord( 'postType', template.type, template.id );
+	}
+
+	return (
+		<>
+			<MenuItem
+				onClick={ () => {
+					setIsModalOpen( true );
+				} }
+			>
+				{ __( 'Rename' ) }
+			</MenuItem>
+			{ isModalOpen && (
+				<Modal
+					title={ __( 'Rename template' ) }
+					closeLabel={ __( 'Close' ) }
+					onRequestClose={ () => {
+						setIsModalOpen( false );
```

Comment:
This seems a little weird to me, as I'd imagine closing the modal would also remove previously edited title. I believe that's the side-effect of using `useEntityProp` on every change? We could create a new state and reset it every time the modal is opened though.

Reply:
Changed to use state.